repo/maintenance/blob_retain.go

```diff
@@ -0,0 +1,97 @@
+package maintenance
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/kopia/kopia/repo"
+	"github.com/kopia/kopia/repo/blob"
+)
+
+// ExtendBlobRetentionTimeOptions provides option for iextending blob retention algorithm.
+type ExtendBlobRetentionTimeOptions struct {
+	Parallel int
+	DryRun   bool
+}
+
+// ExtendBlobRetentionTime extends the retention time of all relevant blobs managed by storage engine with Object Locking enabled.
+func ExtendBlobRetentionTime(ctx context.Context, rep repo.DirectRepositoryWriter, opt ExtendBlobRetentionTimeOptions, safety SafetyParameters) (int, error) {
+	blobCfg := rep.BlobCfg()
+	if !blobCfg.IsRetentionEnabled() || !blobCfg.RetentionExtend {
+		// Blob retention is disabled
+		return 0, nil
+	}
+
+	if opt.Parallel == 0 {
+		opt.Parallel = 16
+	}
+
+	const extendQueueSize = 100
+
+	extend := make(chan blob.Metadata, extendQueueSize)
+
+	var cnt, toExtend uint64
+
+	var eg errgroup.Group
+
+	var prefixes []blob.ID
+
+	extendOpts := blob.ExtendOptions{
+		RetentionMode:   blobCfg.RetentionMode,
+		RetentionPeriod: blobCfg.RetentionPeriod,
+	}
+
+	if !opt.DryRun {
+		// start goroutines to delete blobs as they come.
+		for i := 0; i < opt.Parallel; i++ {
+			eg.Go(func() error {
+				for bm := range extend {
+					if err := rep.BlobStorage().ExtendBlobRetention(ctx, bm.BlobID, extendOpts); err != nil {
+						return errors.Wrapf(err, "unable to extend blob %q", bm.BlobID)
+					}
+					curCnt := atomic.AddUint64(&cnt, 1)
+					if curCnt%100 == 0 {
+						log(ctx).Infof("  extended %v blobs", curCnt)
+					}
+				}
+				return nil
+			})
+		}
+	}
+
+	for _, pfx := range repo.GetLockingStoragePrefixes() {
+		prefixes = append(prefixes, blob.ID(pfx))
+	}
+
+	// iterate unreferenced blobs and count them + optionally send to the channel to be deleted
+	log(ctx).Infof("Extending retention time for blobs...")
+
+	if err := blob.IterateAllPrefixesInParallel(ctx, opt.Parallel, rep.BlobStorage(), prefixes, func(bm blob.Metadata) error {
```

Comment:
i think this logic needs to be slightly more nuanced - we want to extend retention on blobs that are alive, not the ones that happen to be in the repository. Wouldn't this prevent blobs from ever getting dedleted because of delayed blob gc?

Reply:
Unless I'm missing something, I think it is ok as is.  Since garbage collection only 'hides' blobs (which can be done while objects are still under retention), it should be safe to extend the retention on all blobs, even inactive ones.  We do retention extension after gabage-collection to skip extending just-collected blobs, but doing it the other way would only add one additional period to the blob's retention time.